File: src/main/java/dark/leech/text/plugin/security/SecurityConstants.java
package dark.leech.text.plugin.security;

/**
 * Centralized security constants for plugin validation and sandboxing. Provides single source of
 * truth for allowed/blocked APIs and classes.
 */
public final class SecurityConstants {

    private SecurityConstants() {
        // Utility class - prevent instantiation
    }

    // Blocked dangerous class prefixes
    public static final String[] BLOCKED_CLASS_PREFIXES = {
        "java.lang.Runtime",
        "java.lang.ProcessBuilder",
        "java.io.File",
        "java.lang.System",
        "java.lang.reflect",
        "java.net.Socket",
        "java.net.ServerSocket",
        "java.nio.file"
    };

    // Allowed safe class prefixes
    public static final String[] ALLOWED_CLASS_PREFIXES = {
        "java.lang.String",
        "java.lang.Integer",
        "java.lang.Long",
        "java.lang.Double",
        "java.lang.Boolean",
        "java.lang.Object",
        "java.util",
        "java.math"
    };

    // Allowed safe API prefixes for plugin scripts
    public static final String[] ALLOWED_API_PREFIXES = {
        "Http.get",
        "Http.post",
        "Http.request",
        "Html.select",
        "Html.parse",
        "Text.clean",
        "Text.trim",
        "Text.removeTag",
        "Log.add",
        "Util.base64Encode",
        "Util.base64Decode",
        "Util.md5",
        "Util.sha256",
        "Util.sha512",
        "JSON.parse",
        "JSON.stringify"
    };

    /**
     * Check if a class name is allowed.
     *
     * @param className Class name to check
     * @return true if allowed
     */
    public static boolean isClassAllowed(String className) {
        // Block dangerous classes
        for (String blocked : BLOCKED_CLASS_PREFIXES) {
            if (className.startsWith(blocked)) {
                return false;
            }
        }

        // Allow safe classes
        for (String allowed : ALLOWED_CLASS_PREFIXES) {
            if (className.startsWith(allowed)) {
                return true;
            }
        }

        // Default to deny
        return false;
    }

    /**
     * Check if an API call is allowed.
     *
     * @param apiName API name to check
     * @return true if allowed
     */
    public static boolean isApiAllowed(String apiName) {
        if (apiName == null || apiName.isEmpty()) {
            return false;
        }

        for (String allowed : ALLOWED_API_PREFIXES) {
            if (apiName.equals(allowed) || apiName.startsWith(allowed + ".")) {
                return true;
            }
        }
        return false;
    }
}

File: engine/src/commonMain/kotlin/dev/haipham22/leechtext/plugin/security/SecurityConstants.kt
package dev.haipham22.leechtext.plugin.security

/**
 * Constants bảo mật tập trung cho plugin validation/sandboxing (port từ
 * security/SecurityConstants.java). Single source of truth cho API/class được phép.
 */
object SecurityConstants {
    /** Class prefix nguy hiểm — blocked. */
    val BLOCKED_CLASS_PREFIXES =
        arrayOf(
            "java.lang.Runtime",
            "java.lang.ProcessBuilder",
            "java.io.File",
            "java.lang.System",
            "java.lang.reflect",
            "java.net.Socket",
            "java.net.ServerSocket",
            "java.nio.file",
        )

    /** Class prefix an toàn — allowed. */
    val ALLOWED_CLASS_PREFIXES =
        arrayOf(
            "java.lang.String",
            "java.lang.Integer",
            "java.lang.Long",
            "java.lang.Double",
            "java.lang.Boolean",
            "java.lang.Object",
            "java.util",
            "java.math",
        )

    /** API prefix an toàn cho plugin scripts. */
    val ALLOWED_API_PREFIXES =
        arrayOf(
            "Http.get",
            "Http.post",
            "Http.request",
            "Html.select",
            "Html.parse",
            "Text.clean",
            "Text.trim",
            "Text.removeTag",
            "Log.add",
            "Util.base64Encode",
            "Util.base64Decode",
            "Util.md5",
            "Util.sha256",
            "Util.sha512",
            "JSON.parse",
            "JSON.stringify",
        )

    /** Class name có được phép không (block → allow → default deny). */
    fun isClassAllowed(className: String): Boolean {
        // Block class nguy hiểm
        for (blocked in BLOCKED_CLASS_PREFIXES) {
            if (className.startsWith(blocked)) return false
        }
        // Allow class an toàn
        for (allowed in ALLOWED_CLASS_PREFIXES) {
            if (className.startsWith(allowed)) return true
        }
        // Default deny
        return false
    }

    /** API call có được phép không (match exact hoặc prefix + "."). */
    fun isApiAllowed(apiName: String?): Boolean {
        if (apiName.isNullOrEmpty()) return false
        for (allowed in ALLOWED_API_PREFIXES) {
            if (apiName == allowed || apiName.startsWith("$allowed.")) return true
        }
        return false
    }
}
